### PR TITLE
feat(std.os.uefi.protocol): add Serial IO

### LINKED ISSUE #16713: Serial IO Protocol Missing in `std.os.uefi.protocols`
The [UEFI Serial IO Protocol](https://uefi.org/specs/UEFI/2.10/12_Protocols_Console_Support.html?highlight=const%20efi_guid#efi-serial-io-protocol) is currently missing in the standard library. I already have local definitions for it, I could create a PR for this.

### PR DESCRIPTION
**Pull request by MattEttler**
the purpose of this PR is to introduce the Serial IO Protocol interface as described in the [UEFI Specification](https://uefi.org/specs/UEFI/2.10/12_Protocols_Console_Support.html?highlight=const%20efi_guid#serial-i-o-protocol)

closes #16713 

tested by running `zig test lib/std/std.zig --zig-lib-dir lib`

### CODE REVIEW

**Comment by Vexu** [comment]
You need to reference this new file in `lib/std/os/uefi/protocol.zig`.

**Comment by MattEttler** [comment]
Ah- good catch. Fixed.

File: lib/std/os/uefi/protocol/serial_io.zig [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    };
+
+    pub const StopBitsType = enum(u32) {
+        DefaultParity,
+        NoParity,
+        EvenParity,
+        OddParity,
+        MarkParity,
+        SpaceParity,
+    };
```
Comment (Vexu) [inline lib/std/os/uefi/protocol/serial_io.zig:71]: Copy paste error?

File: lib/std/os/uefi/protocol/serial_io.zig [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    _write: *const fn (*const SerialIo, *usize, *anyopaque) callconv(cc) Status,
+    _read: *const fn (*const SerialIo, *usize, *anyopaque) callconv(cc) Status,
+
+    ///Resets the serial device.
+    pub fn reset(self: *const SerialIo) Status {
+        return self._reset(self);
+    }
+
+    ///Sets the baud rate, receive FIFO depth, transmit/receive time out, parity, data bits, and stop bits on a serial device.
+    pub fn set_attribute(self: *const SerialIo, baudRate: u64, receiverFifoDepth: u32, timeout: u32, parity: ParityType, dataBits: u8, stopBits: StopBitsType) Status {
```
Comment (Vexu) [inline lib/std/os/uefi/protocol/serial_io.zig:22]:
```suggestion
    pub fn setAttribute(self: *const SerialIo, baudRate: u64, receiverFifoDepth: u32, timeout: u32, parity: ParityType, dataBits: u8, stopBits: StopBitsType) Status {
```
These should be camelCase.

File: lib/std/os/uefi/protocol/serial_io.zig [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+pub const SerialIo = extern struct {
+    revision: u64,
+    _reset: *const fn (*const SerialIo) callconv(cc) Status,
+    _set_attribute: *const fn (*const SerialIo, u64, u32, u32, ParityType, u8, StopBitsType) callconv(cc) Status,
+    _set_control: *const fn (*const SerialIo, u32) callconv(cc) Status,
+    _get_control: *const fn (*const SerialIo, *u32) callconv(cc) Status,
+    _write: *const fn (*const SerialIo, *usize, *anyopaque) callconv(cc) Status,
+    _read: *const fn (*const SerialIo, *usize, *anyopaque) callconv(cc) Status,
+
```
Comment (Vexu) [inline lib/std/os/uefi/protocol/serial_io.zig:15]: Why aren't the `mode` and `device_type_guid` fields here?

**Review by Vexu — CHANGES_REQUESTED** [review]

File: lib/std/os/uefi/protocol/serial_io.zig (hunk 1 above)
Comment (The-King-of-Toasters) [reply lib/std/os/uefi/protocol/serial_io.zig:71]: Seems so: https://uefi.org/specs/UEFI/2.10/12_Protocols_Console_Support.html#efi-parity-type

File: lib/std/os/uefi/protocol/serial_io.zig (hunk 3 above)
Comment (MattEttler) [reply lib/std/os/uefi/protocol/serial_io.zig]: That was a miss on my part, they should be; I have added them now.

  A brief footnote on `device_type_guid`: in the spec it is described as a pointer that may be NULL though the typedef in the spec does not explicitly label it as Optional. I have made it optional as this mirrors what I see in the EDID Protocol in the standard library today. Let me know if I have made the wrong call here and I will revert it.

**Comment by perillo** [comment]
@MattEttler, I think that doc-comments should have a space between the `///` token and the comment text.

The comments used in this PR should be all invalid, like this:
```zig
///Sets the baud rate, receive FIFO depth, transmit/receive time out, parity, data bits, and stop bits on a serial device
```

**Pull request merged**